### ISSUE TITLE
Implement the cactus group

### ISSUE THREAD

**Issue by tscrim** [sagemath/sage#19594]
<div id="comment:0"></div>

We implement the cactus group *J*<sub>n</sub> (of type A). The cactus group has important applications in category and representation theory.

This group is not available in GAP as far as I can tell.

CC:  @sagetrac-ptingley @bsalisbury1

Component: **group theory**

Keywords: **cactus**

Author: **Travis Scrimshaw**

Branch/Commit: **[public/groups/cactus_group_v2-19594](https://github.com/sagemath/sagetrac-mirror/tree/public/groups/cactus_group_v2-19594) @ [`64608ff`](https://github.com/sagemath/sagetrac-mirror/commit/64608ff6abc5b56fc125f0c23942492cc1c7f3fe)**

Reviewer: **Dima Pasechnik**

_Issue created by migration from https://trac.sagemath.org/ticket/19594_

**Comment by tscrim** [comment]
<div id="comment:41" align="right">comment:41</div>

A solution to the word problem has been explicitly described in https://arxiv.org/abs/2209.08813. (Darij had also come up with (I believe) the same solution, but we never got around to implementing it.)

**Comment by tscrim** [comment]
<div id="comment:43" align="right">comment:43</div>

Here is an updated version that takes advantage of an embedding into a right-angled Coxeter group that was described by extending that of Mostovoy that is also implicit in Yu.

(Side note: I believe there is an error in Yu about there being a semidirect product with the symmetric group; in particular, I don't think the short exact sequence with the pure cactus group splits.)

---
New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/291a16d6aa151cec55373e4c23990019885705c9"><code>291a16d</code></a></td><td><code>Rebased and updated version of the cactus group using an embedding into a RACG for normalization.</code></td></tr></table>

**Comment by tscrim** [comment]
Changed branch from **[public/groups/cactus_group-19594](https://github.com/sagemath/sagetrac-mirror/tree/public/groups/cactus_group-19594)** to **[public/groups/cactus_group_v2-19594](https://github.com/sagemath/sagetrac-mirror/tree/public/groups/cactus_group_v2-19594)**

**Comment by tscrim** [comment]
Changed commit from **[`b354f58`](https://github.com/sagemath/sagetrac-mirror/commit/b354f580ec5437e70b4d92502f09aa6206cb2fc1)** to **[`291a16d`](https://github.com/sagemath/sagetrac-mirror/commit/291a16d6aa151cec55373e4c23990019885705c9)**

**Comment by dimpase** [comment]
<div id="comment:44" align="right">comment:44</div>

this seems to give a bunch of new results on cactus groups:
https://hal-normandie-univ.archives-ouvertes.fr/hal-03778835/document

**Comment by tscrim** [comment]
<div id="comment:45" align="right">comment:45</div>

Good find! That is doing exactly what I implemented, and it also has the language I was missing to describe why the algorithm works. Although I find the writing leaves a lot to be desired...(perhaps this is to be considered a pre-preprint?) Anyways, I will add a citation and do a bit of doc cleanup tomorrow based on that.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`291a16d`](https://github.com/sagemath/sagetrac-mirror/commit/291a16d6aa151cec55373e4c23990019885705c9)** to **[`c4ed295`](https://github.com/sagemath/sagetrac-mirror/commit/c4ed295290d0082ec2fd6546e9447d06b6b86961)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:46"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/c4ed295290d0082ec2fd6546e9447d06b6b86961"><code>c4ed295</code></a></td><td><code>Updating doctests and references; adding generators for pure cactus group.</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`c4ed295`](https://github.com/sagemath/sagetrac-mirror/commit/c4ed295290d0082ec2fd6546e9447d06b6b86961)** to **[`53c1c6b`](https://github.com/sagemath/sagetrac-mirror/commit/53c1c6b392881b2cc2e5174209d2c16a1e1530f9)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:47"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/53c1c6b392881b2cc2e5174209d2c16a1e1530f9"><code>53c1c6b</code></a></td><td><code>Added doctests to gen() method.</code></td></tr></table>

**Comment by tscrim** [comment]
<div id="comment:48" align="right">comment:48</div>

I have added that to the references (I also found it on the arXiv and linked to that), updated the documentation, added a method to compute the generators of the pure cactus group, and added a few more doctests and helper functions. This should now be in good working order.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`53c1c6b`](https://github.com/sagemath/sagetrac-mirror/commit/53c1c6b392881b2cc2e5174209d2c16a1e1530f9)** to **[`5439a9d`](https://github.com/sagemath/sagetrac-mirror/commit/5439a9dc58b604e3ef87fea28623d7b4d06eaa30)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:49"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5b387f1ae1366e1f2c89a04923c3b9b92438ee2c"><code>5b387f1</code></a></td><td><code>Rebased and updated version of the cactus group using an embedding into a RACG for normalization.</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/71edbb7377a5ecef94fd465d6318dc5d11d8765d"><code>71edbb7</code></a></td><td><code>Updating doctests and references; adding generators for pure cactus group.</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5439a9dc58b604e3ef87fea28623d7b4d06eaa30"><code>5439a9d</code></a></td><td><code>Added doctests to gen() method.</code></td></tr></table>

**Comment by dimpase** [comment]
<div id="comment:50" align="right">comment:50</div>

there is a doctest error:

```
File "src/sage/groups/finitely_presented_named.py", line 574, in sage.groups.finitely_presented_named.CactusPresentation
Failed example:
    J3 = groups.presentation.Cactus(3); J3
Exception raised:
    Traceback (most recent call last):
      File "/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.groups.finitely_presented_named.CactusPresentation[0]>", line 1, in <module>
        J3 = groups.presentation.Cactus(Integer(3)); J3
    AttributeError: module 'sage.groups.finitely_presented_catalog' has no attribute 'Cactus'
```

which as far as I understand fixable with

```diff
--- a/src/sage/groups/finitely_presented_catalog.py
+++ b/src/sage/groups/finitely_presented_catalog.py
@@ -23,3 +23,4 @@ from .finitely_presented_named import SymmetricPresentation as Symmetric
 from .finitely_presented_named import QuaternionPresentation as Quaternion
 from .finitely_presented_named import AlternatingPresentation as Alternating
 from .finitely_presented_named import BinaryDihedralPresentation as BinaryDihedral
+from .finitely_presented_named import CactusPresentation as Cactus
```

But then one gets

```
sage -t --long --warn-long 39.5 --random-seed=184245881360667580703156576385378771477 src/sage/groups/finitely_presented_named.py
**********************************************************************
File "src/sage/groups/finitely_presented_named.py", line 574, in sage.groups.finitely_presented_named.CactusPresentation
Failed example:
    J3 = groups.presentation.Cactus(3); J3
Exception raised:
    Traceback (most recent call last):
      File "sage/structure/category_object.pyx", line 839, in sage.structure.category_object.CategoryObject.getattr_from_category
        return self.__cached_methods[name]
    KeyError: '_product_on_gens'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.groups.finitely_presented_named.CactusPresentation[0]>", line 1, in <module>
        J3 = groups.presentation.Cactus(Integer(3)); J3
      File "/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/src/sage/groups/finitely_presented_named.py", line 588, in CactusPresentation
        x,y = G._product_on_gens(key, key2)
      File "sage/structure/category_object.pyx", line 833, in sage.structure.category_object.CategoryObject.__getattr__
        return self.getattr_from_category(name)
      File "sage/structure/category_object.pyx", line 848, in sage.structure.category_object.CategoryObject.getattr_from_category
        attr = getattr_from_other_class(self, cls, name)
      File "sage/cpython/getattr.pyx", line 356, in sage.cpython.getattr.getattr_from_other_class
        raise AttributeError(dummy_error_message)
    AttributeError: 'CactusGroup_with_category' object has no attribute '_product_on_gens'
```

**Comment by dimpase** [comment]
<div id="comment:51" align="right">comment:51</div>

there are doctest errors in 2 more files:

```
File "src/sage/groups/perm_gps/permgroup_named.py", line 342, in sage.groups.perm_gps.permgroup_named.SymmetricGroup._coerce_map_from_
Failed example:
    S5._coerce_map_from_(J3)
Expected:
    True
Got:
    <bound method SymmetricGroup._from_cactus_group_element of Symmetric group of order 5! as a permutation group>
**********************************************************************
File "src/sage/groups/perm_gps/permgroup_named.py", line 345, in sage.groups.perm_gps.permgroup_named.SymmetricGroup._coerce_map_from_
Failed example:
    S2._coerce_map_from_(J3)
Expected:
    False
Got:
    <BLANKLINE>
**********************************************************************
File "src/sage/groups/perm_gps/permgroup_named.py", line 363, in sage.groups.perm_gps.permgroup_named.SymmetricGroup._from_cactus_group_element
Failed example:
    S5._from_cactus_group_element(elt)
Expected nothing
Got:
    (2,3)
```
and

```
File "src/sage/combinat/permutation.py", line 6849, in sage.combinat.permutation.StandardPermutations_n._from_cactus_group_element
Failed example:
    P5._from_cactus_group_element(elt)
Expected nothing
Got:
    [1, 3, 2, 4, 5]
```

**Comment by dimpase** [comment]
Reviewer: **Dima Pasechnik**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:52"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d61d2975dbde845e6b1b04ce33e7c35b92576d63"><code>d61d297</code></a></td><td><code>Fixing some doctests and details.</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`5439a9d`](https://github.com/sagemath/sagetrac-mirror/commit/5439a9dc58b604e3ef87fea28623d7b4d06eaa30)** to **[`d61d297`](https://github.com/sagemath/sagetrac-mirror/commit/d61d2975dbde845e6b1b04ce33e7c35b92576d63)**

**Comment by tscrim** [comment]
<div id="comment:53" align="right">comment:53</div>

Thank you. This should take care of those issues.

**Comment by dimpase** [comment]
<div id="comment:54" align="right">comment:54</div>

In finite group theory J_n (n=[1..4]) is a Janko group. Can you add a remark in the docs, to disambiguate ?

https://en.wikipedia.org/wiki/Janko_group

**Comment by dimpase** [comment]
<div id="comment:55" align="right">comment:55</div>

After this, it would be a positive review from me.

**Comment by tscrim** [comment]
<div id="comment:56" align="right">comment:56</div>

Replying to [Dima Pasechnik](#comment%3A54):
> In finite group theory J_n (n=[1..4]) is a Janko group. Can you add a remark in the docs, to disambiguate ?
> 
> https://en.wikipedia.org/wiki/Janko_group

I am not quite sure what you are looking for. Nearly everywhere in the docs, the notation `J_n` was clearly noted as the cactus group (the exception being in the method `geometric_representation_generators()`'s doc). Are you wanting a message saying something to the effect of "Despite having the same notation, this is not the Janko group."? I would be surprised if having the same notation would cause any confusion.

**Comment by dimpase** [comment]
<div id="comment:57" align="right">comment:57</div>

I'm talking about

```
$ git grep J_2
src/sage/graphs/generators/distance_regular.pyx:    Return the distance-transitive graph with automorphism group `J_2`.
src/sage/graphs/generators/distance_regular.pyx:    G.name("J_2 graph")
src/sage/graphs/generators/distance_regular.pyx:        Grassmann graph J_2(6, 3): Graph on 1395 vertices
src/sage/graphs/generators/smallgraphs.py:    permutation representation of the Janko group `J_2`, as described in version
src/sage/graphs/strongly_regular_db.pyx:    This graph is built from the action of `J_2` on the cosets of a `3.PGL(2,9)`-subgroup.
src/sage/graphs/strongly_regular_db.pyx:    g.name('J_2 on cosets of 3.PGL(2,9)')
```

here J_2 is a group - but a quite different one. How about

```diff
--- a/src/sage/groups/finitely_presented_named.py
+++ b/src/sage/groups/finitely_presented_named.py
@@ -12,7 +12,7 @@ Groups available as finite presentations:
 - Alternating group, `A_n` of order `n!/2` --
   :func:`groups.presentation.Alternating <sage.groups.finitely_presented_named.AlternatingPresentation>`
 
-- Cactus group, `J_n` --
+- Cactus group, `J_n` (not to be confused with Janko groups `J_n`) --
   :func:`groups.presentation.Cactus <sage.groups.finitely_presented_named.CactusPresentation>`
 
 - Cyclic group, `C_n` of order `n` --
```

**Comment by tscrim** [comment]
<div id="comment:58" align="right">comment:58</div>

Replying to [Dima Pasechnik](#comment%3A57):
> I'm talking about

I understand your issue that someone out there might confuse the notation with the standard notation for the Janko groups. (I would actually argue that those other places should make clear what group they are talking about since notation has a high chance of conflicting...)

How about sidestepping this by

```diff
--- a/src/sage/groups/finitely_presented_named.py
+++ b/src/sage/groups/finitely_presented_named.py
@@ -12,7 +12,7 @@ Groups available as finite presentations:
 - Alternating group, `A_n` of order `n!/2` --
   :func:`groups.presentation.Alternating <sage.groups.finitely_presented_named.AlternatingPresentation>`
 
-- Cactus group, `J_n` --
+- the `n`-fruit Cactus group --
   :func:`groups.presentation.Cactus <sage.groups.finitely_presented_named.CactusPresentation>`
 
 - Cyclic group, `C_n` of order `n` --
```

**Comment by dimpase** [comment]
<div id="comment:59" align="right">comment:59</div>

You do use `J_n` in several other docstring places in your patch - but now you propose to hide it in the index. No, this is making docs worse, not better. 

I don't get what you don't like about my suggestion. Your patch introduces ambiguity w.r.t. to the meaning of "group J_n" - then please make sure that it does not confuse readers of the documentation.

**Comment by tscrim** [comment]
<div id="comment:60" align="right">comment:60</div>

There is no ambiguity. The notation clearly is referring to the cactus group every place in the documentation. I do not want to write anything saying “this notation is not the same as X, Y, and Z” anytime there is conflicting notation. This is distracting, a maintenance burden, and more likely to introduce confusion.

**Comment by dimpase** [comment]
<div id="comment:61" align="right">comment:61</div>

J_n is standard notation, as such should be mentioned in the catalog/index.

How about PJ_n, should it be in the catalog too?

**Comment by tscrim** [comment]
<div id="comment:62" align="right">comment:62</div>

Replying to [Dima Pasechnik](#comment%3A61):
> J_n is standard notation, as such should be mentioned in the catalog/index.

None of the groups have their notation (standard or otherwise) mentioned in any of the catalogs AFAICS. The only place notation is mentioned (other than in its own file) is in the `finitely_presented_names.py` module-level doc, where it is following the conventions there. However, I would be fine with a change like this:

```diff
--- a/src/sage/groups/finitely_presented_named.py
+++ b/src/sage/groups/finitely_presented_named.py
@@ -12,7 +12,7 @@ Groups available as finite presentations:
 - Alternating group, `A_n` of order `n!/2` --
   :func:`groups.presentation.Alternating <sage.groups.finitely_presented_named.AlternatingPresentation>`
 
-- Cactus group, `J_n` --
+- the `n`-fruit Cactus group, a standard notation for which is `J_n` --
   :func:`groups.presentation.Cactus <sage.groups.finitely_presented_named.CactusPresentation>`
 
 - Cyclic group, `C_n` of order `n` --
```

> How about PJ_n, should it be in the catalog too?

It is in the main `misc` catalog. Are you referring to the finitely-presented one? There isn't a known presentation for the pure cactus group AFAIK. We can derive one since it is a finite-index subgroup of the cactus group, but it is algorithmic (which I didn't feel motivated to do; plus it should have a more general implementation, something the `gens()` code likely should be generalized to).

**Comment by dimpase** [comment]
<div id="comment:63" align="right">comment:63</div>

ok fine, just change as you proposed in the last comment

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`d61d297`](https://github.com/sagemath/sagetrac-mirror/commit/d61d2975dbde845e6b1b04ce33e7c35b92576d63)** to **[`3c4e316`](https://github.com/sagemath/sagetrac-mirror/commit/3c4e3168156312bf975a63941630027e4b91c392)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:64"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3c4e3168156312bf975a63941630027e4b91c392"><code>3c4e316</code></a></td><td><code>Some details about cactus group notation.</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`3c4e316`](https://github.com/sagemath/sagetrac-mirror/commit/3c4e3168156312bf975a63941630027e4b91c392)** to **[`64608ff`](https://github.com/sagemath/sagetrac-mirror/commit/64608ff6abc5b56fc125f0c23942492cc1c7f3fe)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:65"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/64608ff6abc5b56fc125f0c23942492cc1c7f3fe"><code>64608ff</code></a></td><td><code>Some last little details.</code></td></tr></table>

**Comment by tscrim** [comment]
<div id="comment:66" align="right">comment:66</div>

Done. I also noticed a few other trivial things that I fixed in the last commit.

**Comment by dimpase** [comment]
<div id="comment:67" align="right">comment:67</div>

OK, thanks

**Comment by tscrim** [comment]
<div id="comment:68" align="right">comment:68</div>

Thank you for the review.

**Comment by mkoeppe** [comment]
Merged in 10.0.beta0
